src/engine/TransitivePathImpl.h

```diff
@@ -235,6 +238,201 @@ class TransitivePathImpl : public TransitivePathBase {
+    bool startsWithGraphVariable_;
+    // range state
+    bool finished_{false};
+    // The `resultRange_` yields results of this transitive hull
+    // computation. Values from `TableColumnWithVocab::expandUndef` are
+    // processed one by one. The range and iterator are reinitialised when the
+    // next value from the `enumerateRange_` is yielded. The iterator is
+    // incremented on each iteration of this object.
+    std::optional<NodeWithTargetRange> resultRange_;
+    NodeWithTargetRange::iterator result_{};
+    // The `enumerateRange_` yields `pair<index, ZippedType>` for the current
+    // table column's `startNodes_`.  The range and iterator are
+    // reinitialised when the next table column is yielded. The iterator is
+    // incremented only when the `resultRange_` is exhausted.
+    std::optional<EnumerateRangeType> enumerateRange_;
+    ql::ranges::iterator_t<EnumerateRangeType> enumerateRangeIt_{};
+    // This iterator points to the table columns of `startNodes_` and only
+    // incremented when the `enumerateRange_` is exhausted. The transitive hull
+    // computation is completed when  this iterator reaches the end of
+    // `startNodes_`.
+    ql::ranges::iterator_t<Node> tableColumnIt_{};
+
+   public:
+    TransitiveHullLazyRange(const TransitivePathImpl<T>& parent, const T edges,
+                            LocalVocab edgesVocab, Node startNodes,
+                            TripleComponent start, TripleComponent target,
+                            bool yieldOnce)
+        : parent_(parent),
+          edges_(std::move(edges)),
+          edgesVocab_(std::move(edgesVocab)),
+          startNodes_(std::move(startNodes)),
+          target_(std::move(target)),
+          yieldOnce_(yieldOnce) {
+      // `targetId` is only ever used for comparisons, and never stored in the
+      // result, so we use a separate local vocabulary - targetHelper_.
+      const auto& index = parent_.getIndex();
+      targetId_ = target_.isVariable()
+                      ? std::nullopt
+                      : std::optional{std::move(target_).toValueId(
+                            index.getVocab(), targetHelper_,
+                            index.encodedIriManager())};
+      sameVariableOnBothSides_ =
+          !targetId_.has_value() && parent_.lhs_.value_ == parent_.rhs_.value_;
+      endsWithGraphVariable_ = !targetId_.has_value() &&
+                               parent_.graphVariable_ == target_.getVariable();
+      startsWithGraphVariable_ =
+          start.isVariable() && parent_.graphVariable_ == start.getVariable();
+    }
+
+    std::optional<Id> getTargetId(const Id& startNode,
+                                  const Id& graphId) const {
+      if (sameVariableOnBothSides_) {
+        return startNode;
+      } else if (endsWithGraphVariable_) {
+        return graphId;
+      } else {
+        return targetId_;
+      }
+    };
+
+    ad_utility::LoopControl<NodeWithTargets> process(ZippedType& idPair,
```

Comment:
Please add a comment the explains that this function calls the core logic and runs the actual graph traversal with `findConnectedNodes`